File: 1855.maximum-distance-between-a-pair-of-values.cpp
/*
 * @lc app=leetcode id=1855 lang=cpp
 *
 * [1855] Maximum Distance Between a Pair of Values
 */

// @lc code=start
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxDistance(std::vector<int>& nums1, std::vector<int>& nums2) {
        std::reverse(nums1.begin(), nums1.end());
        std::reverse(nums2.begin(), nums2.end());
        int result = 0;
        int n1 = nums1.size();
        int n2 = nums2.size();
        
        for (int i1 = n1 - 1; i1 >= 0; i1--) {
            int ind = std::lower_bound(nums2.begin(), nums2.end(), nums1[i1]) - nums2.begin();
            if (ind < n2 - n1 + i1 + 1) {
                result = std::max(result, n2 - n1 + i1 - ind);
            }
        }
        
        return result;
    }
};
// @lc code=end

File: 1855.maximum-distance-between-a-pair-of-values.py
#
# @lc app=leetcode id=1855 lang=python3
#
# [1855] Maximum Distance Between a Pair of Values
#

# @lc code=start
import bisect
from typing import List


class Solution:
    '''
    Notice nums1 and nums2 are already sorted, but in reverse order
    If we reverse both nums1 and nums2, take n1 = len(nums1) and n2 = len(nums2), the question becomes:
    For each index i1 = n1 - 1 to 0 in nums1, bisect_left for nums1[i1]  inside nums2
    If the index >= n2 - n1 + i1 + 1, we do not find a match
    If we find index < n2 - n1 + i1 + 1, the maximum distance for the current pair is n2 - n1 - index + i1
    '''
    def maxDistance(self, nums1: List[int], nums2: List[int]) -> int:
        nums1.reverse()
        nums2.reverse()
        result = 0
        n1, n2 = len(nums1), len(nums2)
        for i1 in range(n1 - 1, -1, -1):
            ind = bisect.bisect_left(nums2, nums1[i1])
            if ind < n2 - n1 + i1 + 1:
                result = max(result, n2 - n1 + i1 - ind)
        return result
# @lc code=end

